packages/block-library/src/image/edit.native.js

```diff
@@ -1,75 +1,130 @@
 /**
  * External dependencies
  */
-import { View, Image, TextInput } from 'react-native';
+import React from 'react';
+import { View, Image, TextInput, NativeEventEmitter } from 'react-native';
 import RNReactNativeGutenbergBridge from 'react-native-gutenberg-bridge';
 
+const gutenbergBridgeEvents = new NativeEventEmitter( RNReactNativeGutenbergBridge );
+
 /**
  * Internal dependencies
  */
 import { MediaPlaceholder, RichText, BlockControls } from '@wordpress/editor';
-import { Toolbar, ToolbarButton } from '@wordpress/components';
+import { Toolbar, ToolbarButton, Spinner } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
 
-export default function ImageEdit( props ) {
-	const { attributes, isSelected, setAttributes } = props;
-	const { url, caption } = attributes;
-
-	const onUploadPress = () => {
-		// This method should present an image picker from
-		// the device.
-		//TODO: Implement upload image method.
-	};
-
-	const onMediaLibraryPress = () => {
-		RNReactNativeGutenbergBridge.onMediaLibraryPress( ( mediaUrl ) => {
-			if ( mediaUrl ) {
-				setAttributes( { url: mediaUrl } );
-			}
-		} );
-	};
-
-	if ( ! url ) {
+const mediaUploadStateUploading = 1;
+const mediaUploadStateSucceeded = 2;
+const mediaUploadStateFailed = 3;
+
+export default class ImageEdit extends React.Component {
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			progress: 0,
+		};
+
+		this.mediaUpload = this.mediaUpload.bind( this );
+		this.addMediaUploadListener = this.addMediaUploadListener.bind( this );
+		this.removeMediaUploadListener = this.removeMediaUploadListener.bind( this );
+		this.finishMediaUploading = this.finishMediaUploading.bind( this );
+	}
+
+	mediaUpload( payload ) {
+		if ( payload.state === mediaUploadStateUploading ) {
+			this.setState( { progress: payload.progress } );
+		} else if ( payload.state === mediaUploadStateSucceeded || payload.state === mediaUploadStateFailed ) {
+			this.finishMediaUploading( payload );
+		}
+	}
+
+	finishMediaUploading( payload ) {
+		const { setAttributes } = this.props;
+
+		setAttributes( { url: payload.mediaUrl, id: payload.mediaId } );
+		this.setState( { progress: payload.progress } );
+		this.removeMediaUploadListener( payload.mediaId );
+	}
+
+	addMediaUploadListener( mediaId ) {
+		gutenbergBridgeEvents.addListener( 'mediaUpload' + mediaId, this.mediaUpload );
```

Comment:
I believe we will have a problem here with iOS. As you can see [here](
https://facebook.github.io/react-native/docs/native-modules-ios#sending-events-to-javascript), on iOS we need to know ahead of time what events do we support by overriding `- (NSArray<NSString *> *)supportedEvents` and returning an array with the event names.

Because of that, we can not have dynamic events names like appending the media id to it.

Do you think it would be possible to send the media id as a parameter instead?

Reply:
> Do you think it would be possible to send the media id as a parameter instead?

Good call there 👍

Reply:
Hey @etoledom, thanks for the update. I wasn't aware of this one as this approach was working in some lower version. I will fix this. Thanks for letting me know.

cc: @SergioEstevao